cameraview/src/main/java/com/otaliastudios/cameraview/engine/Camera2Engine.java

```diff
@@ -902,8 +902,44 @@ public void setPlaySounds(boolean playSounds) {
     }
 
     @Override
-    public void setZoom(float zoom, @Nullable PointF[] points, boolean notify) {
-        // TODO
+    public void setZoom(final float zoom, final @Nullable PointF[] points, final boolean notify) {
+        mHandler.run(new Runnable() {
+            @Override
+            public void run() {
+                if (!mCameraOptions.isZoomSupported()) return;
+
+                Rect newRect = getZoomRect(zoom);
+                if(newRect != null){
+                    mRepeatingRequestBuilder.set(CaptureRequest.SCALER_CROP_REGION, newRect);
+                    applyRepeatingRequestBuilder();
+                    mZoomValue = zoom;
+
+                    if (notify) {
+                        mCallback.dispatchOnZoomChanged(zoom, points);
+                    }
+                }
+            }
+        });
+    }
+
+    private Rect getZoomRect(float zoomLevel) {
```

Comment:
I don't know what's happening here but I trust you :-) this said I think you can make this return @NonNull always because the incoming zoom value is always between 0 and 1 so it's valid?

Reply:
done.